### PR TITLE
fix: read url credentials with the WHATWG parser

### PR DESCRIPTION
**Pull request by luantaraschi**
﻿`extractAuthFromUrl` splits the userinfo with `userpass.split(':')` and destructures the first two parts, so everything after a second colon is discarded:

```js
extractAuthFromUrl('https://user:pa:ss@github.com/owner/repo.git')
// { url: 'https://github.com/owner/repo.git',
//   auth: { username: 'user', password: 'pa' } }
```

Only the first colon separates the two halves. The WHATWG URL parser reads the same string as user `user` and password `pa:ss`:

```js
new URL('https://user:pa:ss@h/r').password   // 'pa%3Ass'
```

`GitRemoteHTTP` hands that object straight to `calculateBasicAuthHeader`, so the request goes out with a truncated secret. The server answers 401 and nothing in the message hints that the password was cut, which makes it a slow thing to diagnose from the outside.

## The change

Split on the first colon. A userinfo with no colon still gives an `undefined` password, so the username-only path keeps behaving as it does today.

## Tests

`__tests__/test-utils-extractAuthFromUrl.js`, six cases: no credentials, plain user and password, a colon inside the password, username with no password, empty password, and a cross-check that the username and password match what `new URL()` reads for the same input.

Two of the six fail on `main` and pass here; the other four pass on both sides, so a harness that stopped reaching the function would not look like a fix.

`npx jest __tests__/test-utils-extractAuthFromUrl.js` with `TEST_ENV=node`: 6 passed. `npx eslint src __tests__` is clean.

## One note

The test reaches the function through `isomorphic-git/internal-apis`, following the precedent of `utils/join`, so this adds one export line to `src/internal-apis.js`. My two other open PRs (#2397, #2398) add a line to the same alphabetical list, so whichever lands last will want a trivial rebase. Say the word and I will rebase this one whenever it suits the order you prefer.

## Summary by CodeRabbit

* **New Features**
  * Exposed URL credential extraction through the public API.
  * Added support for case-insensitive HTTP(S) schemes and percent-decoded credentials.

* **Bug Fixes**
  * Preserves additional colons in passwords.
  * Returns no password when credentials contain only a username.
  * Handles invalid percent-encoding safely while preserving the original URL.
  * Ignores `@` characters in query strings and fragments.

* **Tests**
  * Added comprehensive coverage for credential extraction, decoding, edge cases, and standard URL parsing compatibility.

### CODE REVIEW

**Comment by jcubic** [comment]
Maybe it's better to check if the URL exists and use it directly. You already use it in tests.

The URL in browsers is in the baseline: https://caniuse.com/url

But we can keep the old code as a fallback.

**Comment by jcubic** [comment]
Obviously, passwords with just `%` be encoded by URL as `%25`.

**Comment by luantaraschi** [comment]
I took your suggestion and switched the credential parsing to `URL` in `668b7d2e`.

The parser is used only to read the username and password. The function still removes credentials from the original string, so it does not normalize host case, remove `:443`, or add a trailing slash. `decodeURIComponent` is guarded so a bare `%` does not throw, and the previous split remains as a fallback when `URL` is unavailable or rejects the input.

The tests cover password colons, uppercase schemes, percent-decoding, malformed escapes, empty or missing credentials, and preservation of the original URL shape. Follow-up commit `2dabc35` also stops the credential match at `/`, `?`, and `#`, with regressions for query and fragment values containing `@`.

Ten tests pass. This replaces the initial colon-only fix rather than stacking another parser on top of it.

File: src/utils/extractAuthFromUrl.js [hunk 1]
```diff
@@ -1,16 +1,69 @@
+// The `@` that ends the credentials is the last one before the path, and the
+// scheme is matched case-insensitively: `HTTPS://user:pass@host/x` is a URL
+// like any other, and leaving the credentials in it is what `fetch` rejects.
+const CREDENTIALS = /^(https?:\/\/)([^/]+)@/i
```
[the remark this answers is not in the document]
Comment (luantaraschi) [reply src/utils/extractAuthFromUrl.js]: Checked it, and it holds. `https://github.com?account=user@example.com` came back as `https://example.com`, so the request would have gone out against a different host than the one written.

Two things worth recording about the scope. It predates this PR: `main` matches `([^/]+)@` as well and rewrites the same string the same way. And it needs a url with no path segment, because a `/` already ends the match, so `https://github.com/o/r.git?token=a@b` was never affected. A remote written the usual way does not reach it.

Fixed in `2dabc35` by ending the authority at the first `/`, `?` or `#`, with a test covering both the query and the fragment. That test fails on the previous expression and passes on this one.

On the empty userinfo: `https://@github.com/o/r.git` still comes back with the url untouched and `auth: {}`, exactly as before, since `[^/?#]+` does not match there either. I left that as it is rather than widening the expression, since allowing it through would change what the function returns rather than fix anything.

**Review by jcubic — APPROVED** [review]
Looks good.
